src/apps/competitions/models.py

```diff
@@ -23,6 +28,107 @@ def __str__(self):
     def bundle_dataset(self):
         return CompetitionCreationTaskStatus.objects.get(resulting_competition=self).dataset
 
+    def check_future_phase_submissions(self):
+        """
+        Checks for if we need to migrate current phase submissions to next phase.
+        """
+        phases = self.phases.all()
+
+        if len(phases) == 0:
+            return
+
+        try:
+            current_phase = phases.get(status='Current')
+            next_phase = phases.get(status='Next')
+            final_phase = phases.get(status='Final')
+
+        except Phase.DoesNotExist:
+            return
+
+        # Making sure current_phase or next_phase is not None
+        if current_phase is None or next_phase is None or current_phase is final_phase:
+            return
+
+        logger.info("Checking for needed migrations on competition pk=%s, current phase: %s, next phase: %s" %
+                    (self.pk, current_phase.index, next_phase.index))
+
+        # Check for next phase and see if it has auto_migration enabled
+        if next_phase and next_phase.auto_migration:
+            if next_phase.start < now() and current_phase.end < now():
+                self.apply_phase_migration(current_phase, next_phase)
+
+    def apply_phase_migration(self, current_phase, next_phase):
+        '''
+        Does the actual migrating of submissions from current_phase to next_phase
+
+        :param current_phase: The phase object to transfer submissions from
+        :param next_phase: The new phase object we are entering
+        '''
+        logger.info("Checking for submissions that may still be running competition pk=%s" % self.pk)
+
+        if current_phase.submissions.filter(status='Scoring').exists():
+            logger.info('Some submissions still marked as processing for competition pk=%s' % self.pk)
+            self.is_migrating_delayed = True
+            self.save()
+            return
+        else:
+            logger.info("No submissions running for competition pk=%s" % self.pk)
+
+        logger.info('Doing phase migration on competition pk=%s from phase: %s to phase: %s' %
+                    (self.pk, current_phase.index, next_phase.index))
+
+        if self.is_migrating:
+            logger.info('Trying to migrate competition pk=%s, but it is already being migrated!' % self.pk)
+            return
+
+        self.is_migrating = True
+        self.save()
+
+        try:
+            submission_list = []
+            submissions = Submission.objects.filter(phase=current_phase)
+
+            for submission in submissions:
+                submission_list.append(submission)
+
+            participants = {}
+
+            for s in submission_list:
+                if s.is_migrated is False:
+                    participants[s.participant] = s
```

Comment:
If I understand this logic right, this only migrates one submission per user each time this runs, because the participant object is the same for all the user's submissions, so it just keeps overwriting the same key. Is this the desired functionality?

Reply:
can we do something like: 
```python
for submission in Submission.objects.filter(phase=current_phase, is_migrated=False):
    new_submission = Submission(
        participant=submission.participant,
        # and all the other fields
    )
    new_submission.save(ignore_submission_limit=True)
    new_submission.start()

    submission.is_migrated = True
    submission.save()
```
Then we do this more than one submission per participant at a time. (unless that is the desired functionality, in which case I have some ideas for cool nested queries so you can still just loop a single queryset)

Reply:
replaced the logic with what you've got and it looks much cleaner! Thanks Jimmy!